File: OpenMP/char_freq/char_freq_seq.c
#include <stdio.h> 
#include <stdlib.h> 
#include <time.h>
#define N 128
#define base 0

void zeros(int array[], int n);
void count_characters(int freq[], char buffer[], long file_size);
void display_count(int freq[], int n);

int main (int argc, char *argv[]) {
	
	FILE *pFile;
	long file_size;
	char * buffer;
	char * filename;
	size_t result;
	int freq[N];
	clock_t start, end;

    if (argc != 2) {
		printf ("Usage : %s <file_name>\n", argv[0]);
		return 1;
    }
	filename = argv[1];
	pFile = fopen ( filename , "rb" );
	if (pFile==NULL) {printf ("File error\n"); return 2;}

	// obtain file size:
	fseek (pFile , 0 , SEEK_END);
	file_size = ftell (pFile);
	rewind (pFile);
	printf("file size is %ld\n", file_size);
	
	// allocate memory to contain the file:
	buffer = (char*) malloc (sizeof(char)*file_size);
	if (buffer == NULL) {printf ("Memory error\n"); return 3;}

	// copy the file into the buffer:
	result = fread (buffer,1,file_size,pFile);
	if (result != file_size) {printf ("Reading error\n"); return 4;} 

	zeros(freq, N);

    start = clock();

	count_characters(freq, buffer, file_size);

	end = clock();

	display_count(freq, N);	

	(void) printf("Time spent for counting: %g", (double)(end-start));

	fclose (pFile);
	free (buffer);

	return 0;
}

/*
 * Function:  zeros 
 * --------------------
 * Initializes an integer array with zeros
 *
 *  array: the array that will be filled with zeros
 *  n: number of elements in the array
 *
 */

void zeros(int array[], int n){
	int j;
	for (j=0; j<n; ++j)
		array[j]=0;
}

/*
 * Function:  count_characters 
 * --------------------
 * Counts the frequency of characters in an char array
 *
 *  freq: the array that will contain each character's frequency
 *  buffer: the array that contains the characters
 *  file_size: size of buffer
 *
 */

void count_characters(int freq[], char buffer[], long file_size){
	int i;
	for (i=0; i<file_size; ++i)
		++freq[buffer[i] - base];
}

/*
 * Function:  display count 
 * --------------------
 * Prints the array of characters' frequency
 *
 *  freq: the array of frequencies
 *  n: number of elements in the freq array
 *
 */

void display_count(int freq[], int n){
	int j;
	for (j=0; j<n; ++j)
		(void) printf("%d = %d\n", j+base, freq[j]);
}

File: MPI/mpi4py/char_freq/char_freq_seq.py

import sys
from os import SEEK_END
import numpy as np
from mpi4py import MPI

NUM_OF_ARGS = 2
N = 128
base = 0

def check_args():
    '''
    Function:  check_args 
    --------------------
    Checks if the number of arguments is correct
    '''
    # if no filename given, exit
    if len(sys.argv) != NUM_OF_ARGS:
        print(f'Usage : {sys.argv[0]} <file_name>\n')
        exit(1)
    return sys.argv[1]

def open_file(mode='r'):
    '''
    Function:  open_file 
    --------------------
    Opens input file size
    mode: method (mode) for opening the file
    '''
    try:
        f = open(filename, mode)
    except Exception as e:
        print(e)
        exit(2)
    return f

def obtain_file_size(f):
    '''
    Function:  obtain_file_size 
    --------------------
    Obtains input file's size
    f: input file
    '''
    f.seek(0, SEEK_END)
    file_size = f.tell()
    rewind(f)
    return file_size

def count_characters(freq, buffer, file_size):
    '''
    Function:  count_characters 
    --------------------
    Counts the frequency of characters in an char array
    
    freq: the array that will contain each character's frequency
    buffer: the array that contains the characters
    file_size: size of buffer
    '''
    for i in range(file_size):
    	freq[ord(buffer[i]) - base] += 1

def rewind(f):
    '''
    Function:  rewind 
    --------------------
    Set the file position to the beginning of the file 
    of the given stream

    f: input file
    '''
    f.seek(0)

if __name__ == '__main__':

    # check arguments, if file name is given return it
    filename = check_args()

    # open file (default mode='r')
    f = open_file()

    # get file size
    file_size = obtain_file_size(f)
    
    print(f'File size is {file_size}\n')

	# copy the file into the buffer
    buffer = f.read(file_size)

    # frequency vector init with zeros
    freq = np.zeros(N, dtype=np.int)

    # start counting
    start = MPI.Wtime()

    # count characters of buff (of file)
    count_characters(freq, buffer, file_size)

    # done
    end = MPI.Wtime()

    # print char frequency and time of execution
    [print(f'{j+base} = {freq[j]}') for j in range(N)]
    print(f'Time spent for counting: {end-start}')

    # close file
    f.close()
